lib/spack/llnl/util/lang.py

```diff
@@ -193,15 +193,19 @@ def memoized(func):
     func.cache = {}
 
     @functools.wraps(func)
-    def _memoized_function(*args):
-        if not isinstance(args, Hashable):
+    def _memoized_function(*args, **kwargs):
+        assert isinstance(args, tuple), args
+        key = args + tuple(kwargs.items() if kwargs else [])
```

Comment:
I think this generalization may introduce unexpected behavior. If I'm reading this correctly a function signature like:
```python
def foo(bar, baz):
    pass
```
would generate three different keys for these calls:
```python
foo(1, 2)  # key should be (1, 2)
foo(1, baz=2)  # key should be (1, ('baz, 2))
foo(bar=1, baz=2)  # key should be (('bar', 1), ('baz', 2))
```
which may be associated with three different values depending on the call. 

Also: we may extract this small change, if needed, into a small PR of its own.

Reply:
Yes! Thank you. It has been reverted.

Reply:
There is a battle-tested implementation in `memo.py` in pants: https://github.com/pantsbuild/pants/blob/master/src/python/pants/util/memo.py I would borrow from that if we were to make another PR for this.